### PR TITLE
 [Type checker] Prefer available declarations in AnyObject lookup.

### PR DESCRIPTION
**Pull request by DougGregor**
When we perform AnyObject lookup, we can find multiple declarations with
the same signature. When we do so, we pick fairly arbitrarily among them.
However, make sure that we don’t pick an unavailable declaration when
an available declaration is… available.

Fixes rdar://problem/39115471.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ImportType.cpp [hunk 1]
```diff
@@ -2015,9 +2015,6 @@ ImportedType ClangImporter::Implementation::importMethodType(
       continue;
     }
 
-    if (kind == SpecialMethodKind::NSDictionarySubscriptGetter)
-      nonNullArgs.empty();
```
Comment (jrose-apple) [inline lib/ClangImporter/ImportType.cpp]: Oh no. :-( RIP, code that was trying to call `clear()` but never did.

**Review by jrose-apple — APPROVED** [review]
LGTM \*cry\*

File: lib/ClangImporter/ImportType.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/ClangImporter/ImportType.cpp]: Heh, yeah. This isn't actually needed for this PR, but happened to be in my local tree.

Comment (DougGregor) [reply lib/ClangImporter/ImportType.cpp]: Too late to enable it now ;)

**Review by xedin — APPROVED** [review]
LGTM! I curious why we filter results on `AnyObject` this way? Shouldn't different overloads with the same type still form a solution but then be ranked based on attributes?

**Comment by DougGregor** [comment]
We wouldn't be able to rank them (i.e., it would almost always be ambiguous), because we're likely to have the same kind of declaration in many different classes/protocols. That's actually okay--we have no idea, statically, which method `objc_msgSend` will dynamically call--so "just pick one".

We could delay this "just pick one" until later in type checker, I guess, treating it as an overload set and somehow collapsing the results later, but I'd be concerned about increasing the size of overload sets considerably for little expected gain.

**Pull request merged**